File: sdk/java/src/main/java/com/anyvali/schemas/OptionalSchema.java
package com.anyvali.schemas;

import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Wraps a schema to make it optional (accepts absent/undefined values).
 */
public class OptionalSchema extends Schema<Object> {
    private final Schema<?> inner;

    public OptionalSchema(Schema<?> inner) {
        this.inner = inner;
        // Inherit defaults from inner
        if (inner.hasDefaultValue()) {
            this.hasDefault = true;
            this.defaultValue = inner.getDefaultValue();
        }
    }

    private OptionalSchema(OptionalSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.inner = other.inner;
    }

    @Override
    protected boolean acceptsNull() {
        return inner.acceptsNullValue();
    }

    @Override
    public Object runPipeline(Object input, ValidationContext ctx) {
        // If absent (sentinel), return null without error
        if (input == ABSENT) {
            if (hasDefault) {
                return deepCopyValue(defaultValue);
            }
            return null;
        }
        return inner.runPipeline(input, ctx);
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        return inner.validateValue(input, ctx);
    }

    @Override
    protected Map<String, Object> toNode() {
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", "optional");
        node.put("schema", inner.toPortableNode());
        return addCommonNodeFields(node);
    }

    @Override
    protected Schema<Object> copy() {
        return new OptionalSchema(this);
    }

    public Schema<?> getInner() {
        return inner;
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/OptionalSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

data class OptionalSchema(
    val inner: Schema<*>,
    val defaultValue: Any? = UNSET
) : Schema<Any?>() {
    override val kind: String = "optional"

    fun default(v: Any?) = copy(defaultValue = v)

    fun describe(description: String, opts: DescribeOptions? = null): OptionalSchema {
        applyDescribe(description, opts)
        return this
    }

    fun metadata(meta: Map<String, Any?>, replace: Boolean = false): OptionalSchema {
        applyMetadata(meta, replace)
        return this
    }

    override fun safeParseWithContext(input: Any?, ctx: ValidationContext): ParseResult<Any?> {
        // OptionalSchema should not be called directly with absent values;
        // absent handling is done in ObjectSchema.
        // If called, it delegates to inner.
        return inner.safeParseWithContext(input, ctx)
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        return inner.validateValue(value, ctx)
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("optional"))
        put("schema", inner.exportNode())
        if (defaultValue !== UNSET) {
            when (defaultValue) {
                null -> put("default", JsonNull)
                is String -> put("default", JsonPrimitive(defaultValue))
                is Boolean -> put("default", JsonPrimitive(defaultValue))
                is Number -> put("default", JsonPrimitive(defaultValue))
                else -> put("default", JsonPrimitive(defaultValue.toString()))
            }
        }
        addMetadataToNode(this)
    }

    companion object {
        val UNSET = Any()
    }
}
